File: Examples/Simulation_C/Simulation_4C.m
% Multivariate Bayesian Inversion for Classification and Regression
% Simulation 4: comparison of methods (MATLAB script)
% 
% Author: Joram Soch, OvGU Magdeburg
% E-Mail: joram.soch@ovgu.de
% 
% Version History:
% - 20/02/2022, 22:38: first MATLAB version
% - 20/02/2025, 16:48: final MATLAB version
% - 26/02/2026, 13:02: added performance plot
% - 26/02/2026, 14:02: rewrote MBR and SVR
% - 26/02/2026, 14:32: added MLR, RFR, NNR
% - 26/02/2026, 15:32: added GNB regression
% - 26/02/2026, 15:44: refined performance plot


clear
close all

%%% Step 1: specify ground truth %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% set ground truth
rng(2);
n  = 200;                       % number of data points
k  = 10;                        % number of CV folds
v  = 10;                        % number of features
mu = 0;                         % beta mean
sb = 1;                         % beta variance
s2 = 1;                         % noise variance
tau= 0.25;                      % time constant
V  = toeplitz(tau.^[0:1:(n-1)]);% temporal covariance
ny = 0.5;                       % space constant
Si = toeplitz( ny.^[0:1:(v-1)]);% spatial covariance

% generate targets
xm = 1;                         % -1 < x < +1
x  = (2*xm)*rand(n,1)-xm;       % x ~ U(-1,+1)
X  = [x, ones(size(x))];        % design matrix


%%% Step 2: generate data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% generate data
B = normrnd(mu, sqrt(sb), [size(X,2) v]);
E = matnrnd(zeros(n,v), s2*V, Si, 1);
Y = X*B + E;


%%% Step 3: analyze data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% specify analysis parameters
meth    = {'MBR', 'GNB', 'MLR', 'SVR', 'RFR', 'NNR'};
prior.x = [-xm:0.01:+xm];       % MBR prior distribution
prior.p = (1/range(prior.x))*ones(size(prior.x));
Dgnb    = 'mvn';                % GNB distribution name
Mmlr    = 'WLS';                % MLR estimation method
Csvm    = 1;                    % SVM cost parameter
Mrf     = 'Bag';                % RF aggregation method
Lnn     = [10, 10];             % NN hidden layer sizes
Ann     = 'none';               % NN activation function
CV      = ML_CV(n, k, 'kf');    % k-fold cross-validation

% preallocate results
M   = numel(meth);
xp  = zeros(n,M);
r   = zeros(1,M);
MAE = zeros(1,M);

% perform regression
for g = 1:size(CV,2)

    % get training and test set
    i1 = find(CV(:,g)==1);
    i2 = find(CV(:,g)==2);
    Y1 = Y(i1,:);
    Y2 = Y(i2,:);
    x1 = x(i1);
    x2 = x(i2);
    X1 = X(i1,:);
    X2 = X(i2,:);
    V1 = V(i1,i1);
    V2 = V(i2,i2);

    % evaluate all methods
    for h = 1:M
        
        % Method: multivariate Bayesian regression (MBI)
        % https://github.com/JoramSoch/MBI/blob/main/MATLAB/ML_MBI.m
        if strcmp(meth{h}, 'MBR')
            mba1 = mbitrain(Y1, x1, [], V1, 'MBR');
            pp2  = mbitest(Y2, x2, [], V2, mba1, prior);
            [m,j]= max(pp2, [], 2);
            xp2  = prior.x(j);
        end;
        
        % Method: Gaussian naive Bayes (custom)
        % https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Probabilistic_model
        if strcmp(meth{h}, 'GNB')
            B_est1  = (X1'*X1)^-1 * X1'*Y1;
            s2_est1 = (1/n) * sum((Y1-X1*B_est1).^2);
            Si_est1 = diag(s2_est1);
            log_PP  = zeros(numel(i2),numel(prior.x));
            pp2     = zeros(numel(i2),numel(prior.x));
            for i = 1:numel(i2)
                y2i = Y2(i,:);
                for j = 1:numel(prior.x)
                    x2ij        = [prior.x(j), 1];
                    log_PP(i,j) = log(mvnpdf(y2i, x2ij*B_est1, Si_est1)) + ...
                                  log(prior.p(j));
                end;
                pp2(i,:) = exp(log_PP(i,:) - mean(log_PP(i,:)));
                pp2(i,:) = pp2(i,:)./trapz(prior.x, pp2(i,:));
            end;
            [m,j]= max(pp2, [], 2);
            xp2  = prior.x(j);
        end;
        
        % Method: multiple linear regression (MACS)
        % https://github.com/JoramSoch/MACS/blob/master/ME_GLM.m
        if strcmp(meth{h}, 'MLR')
            P1  = inv(V1);
            if strcmp(Mmlr, 'OLS')
                b_est1 = (Y1'*Y1)^-1 * Y1'*x1;
            elseif strcmp(Mmlr, 'WLS')
                b_est1 = (Y1'*P1*Y1)^-1 * Y1'*P1*x1;
            else
                b_est1 = zeros(v,1);
            end;
            xp2 = Y2 * b_est1;
        end;
        
        % Method: support vector regression (LibSVM)
        % https://github.com/JoramSoch/ML4ML/blob/main/ML_SVR.m
        if strcmp(meth{h}, 'SVR')
            opt  = sprintf('-s 4 -t 0 -c %s -q', num2str(Csvm));
            svm1 = svmtrain(x1, Y1, opt);
            xp2  = svmpredict(x2, Y2, svm1, '-q');
        end;

        % Method: random forrest regression (MATLAB)
        % https://de.mathworks.com/help/stats/select-predictors-for-random-forests.html
        if strcmp(meth{h}, 'RFR')
            rfe1 = fitrensemble(Y1, x1, 'Method', Mrf);
            xp2  = predict(rfe1, Y2);
        end;

        % Method: neural network regression (MATLAB)
        % https://de.mathworks.com/help/stats/classificationneuralnetwork.html
        if strcmp(meth{h}, 'NNR')
            nn1 = fitrnet(Y1, x1, 'LayerSizes', Lnn, 'Activations', Ann);
            xp2 = predict(nn1, Y2);
        end;
        
        % store test set predictions
        xp(i2,h) = xp2;
        
    end;
    clear mba1 B_est1 s2_est Si_est1 b_est1 svm1 rfe1 nn1
    
end;

% calculate performance
for h = 1:M
    r(h)   = corr(xp(:,h), x);
    MAE(h) = mean(abs(xp(:,h)-x));
end;


%%% Step 4: visualize results %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% open figure
figure('Name', 'Simulation 4 (Comparison)', 'Color', [1 1 1], 'Position', [50 50 800 900]);
cols  = 'bcmryw';
x_off = 2;
y_off = 0.925;
hold on;

% plot performance
for h = 1:M
    bar(h, r(h), 0.7, cols(h));
end;
plot([(1-1), (M+2)], [0, 0], ':k', 'LineWidth', 2);
plot([x_off, x_off]+1/2, [0, 1], '-k', 'LineWidth', 1);
axis([(1-1), (M+2), 0, 1]);
set(gca,'Box','On');
set(gca,'XTick',[1:M],'XTickLabel',meth);
legend([meth, {'chance'}], 'Location', 'NorthEast');
xlabel('regression approach', 'FontSize', 16);
ylabel('predictive correlation', 'FontSize', 16);
title('Simulation 4: Comparison', 'FontSize', 24);
text(x_off+1/2, y_off, sprintf('generative   \nmethods   '), ...
     'FontSize', 16, 'HorizontalAlignment', 'right', 'VerticalAlignment', 'middle');
text(x_off+1/2, y_off, sprintf('   discriminative\n   methods'), ...
     'FontSize', 16, 'HorizontalAlignment', 'left', 'VerticalAlignment', 'middle');

File: Examples/Simulation_C/Simulation_4C.py
# -*- coding: utf-8 -*-
"""
Multivariate Bayesian Inversion for Classification and Regression
Simulation 4: comparison of methods (Python script)

Author: Joram Soch, OvGU Magdeburg
E-Mail: joram.soch@ovgu.de

Version History:
- 2022-02-20, 22:38: first version in MATLAB
- 2025-03-05, 11:00: final version in Python
- 2026-06-17, 15:46: rewrote MBR and SVR,
                     implemented RGA measure
- 2026-06-18, 13:34: added LinReg, RFC, NNC
- 2026-06-18, 20:15: added GNB regression
"""


# specify MBI path
MBI_dir = '../../Python/'

# import packages
import os
import numpy as np
import scipy.linalg as sp_linalg
import scipy.stats as sp_stats
import sklearn as skl
import matplotlib.pyplot as plt

# import MBI module
orig_dir = os.getcwd()
os.chdir(MBI_dir)
import MBI
os.chdir(orig_dir)


### Step 1: specify ground truth ##############################################

# set ground truth
np.random.seed(1)
n  = 200                        # number of data points
k  = 10                         # number of CV folds
v  = 10                         # number of features
mu = 0                          # beta mean
sb = 1                          # beta variance
s2 = 1                          # noise variance
tau= 0.25                       # time constant and temporal covariance
V  = sp_linalg.toeplitz(np.power(tau, np.linspace(0,n-1,n)))
ny = 0.5                        # space constant and spatial covariance
Si = sp_linalg.toeplitz(np.power( ny, np.linspace(0,v-1,v)))

# generate targets
xm = 1                          # -1 < x < +1, x ~ U(-1,+1)
x  = np.random.uniform(-xm, +xm, size=(n,1))
X  = np.c_[x, np.ones(x.shape)] # design matrix
x  = np.squeeze(x)

# histogram of targets
lim = 1.5
dx  = 0.1
nx  = int((2*lim)/dx)+1
xb  = np.linspace(-lim, +lim, nx)
nb  = np.histogram(x, bins=xb)[0]


### Step 2: generate data #####################################################

# generate data
B = np.random.normal(mu, np.sqrt(sb), size=(X.shape[1],v))
E = MBI.matnrnd(np.zeros((n,v)), s2*V, Si, 1)
Y = X @ B + E


### Step 3: analyze data ######################################################

# specify analysis parameters
meth    = ['GNB', 'MBR', 'LinReg', 'SVR', 'RFR', 'NNR']
prior   = {}
prior['x'] = np.arange(-1, +1+0.01, 0.01)   # MBR prior distribution
prior['p'] = (1/(np.max(prior['x'])-np.min(prior['x'])))*np.ones(prior['x'].size)
Dgnb    = 'mvn'                             # GNB distribution name
Mlinreg = 'WLS'                             # LinReg estimation method
Ksvm    = 'linear'                          # SVM kernel type
Csvm    =  1                                # SVM cost parameter
Nrf     = 100                               # RF number of trees
Lnn     = (10, 10)                          # NN hidden layer sizes
Ann     = 'identity'                        # NN activation function

# specify cross-validation
MBR = MBI.cvMBI(Y, x, V=V, mb_type='MBR')
MBR.crossval(k=k, cv_mode='kf')             # k-fold cross-validation
CV  = MBR.CV

# preallocate results
M   = len(meth)
xp  = np.zeros((n,M))
r   = np.zeros(M)
MAE = np.zeros(M)
RGA = np.zeros(M)

# perform classification
for g in range(k):

    # get training and test set
    i1 = np.array(np.nonzero(CV[:,g]==1)[0], dtype=int)
    i2 = np.array(np.nonzero(CV[:,g]==2)[0], dtype=int)
    Y1 = Y[i1,:]
    Y2 = Y[i2,:]
    x1 = x[i1]
    x2 = x[i2]
    X1 = X[i1,:]
    X2 = X[i2,:]
    V1 = V[i1,:][:,i1]
    V2 = V[i2,:][:,i2]
    
    # evaluate all methods
    for h in range(M):
        
        # Method: multivariate Bayesian regression (MBI)
        # https://github.com/JoramSoch/MBI/blob/main/Python/MBI.py
        if meth[h] == 'MBR':
            mba1 = MBI.model(Y1, x1, V=V1, mb_type='MBR').train()
            pp2  = MBI.model(Y2, x2, V=V2, mb_type='MBR').test(mba1, prior)
            ip2  = np.argmax(pp2, axis=1)
            xp2  = prior['x'][ip2]
        
        # Method: Gaussian naive Bayes (custom)
        # https://en.wikipedia.org/wiki/Naive_Bayes_classifier#Probabilistic_model
        if meth[h] == 'GNB':
            B_est1  = np.linalg.inv(X1.T @ X1) @ X1.T @ Y1
            s2_est1 = (1/n) * np.sum(np.power(Y1 - X1 @ B_est1, 2), axis=0)
            Si_est1 = np.diag(s2_est1)
            log_PP  = np.zeros((i2.size, prior['x'].size))
            pp2     = np.zeros((i2.size, prior['x'].size))
            for i in range(i2.size):
                y2i = np.array([ Y2[i,:] ])
                for j in range(prior['x'].size):
                    x2ij        = np.array([ [prior['x'][j], 1] ])
                    ptd         = sp_stats.multivariate_normal(
                                      (x2ij @ B_est1)[0,:], Si_est1)
                    log_PP[i,j] = np.log(ptd.pdf(y2i))
                pp2[i,:] = np.exp(log_PP[i,:] - np.mean(log_PP[i,:]))
                pp2[i,:] = pp2[i,:] / np.trapz(pp2[i,:], prior['x'])
            ip2     = np.argmax(pp2, axis=1)
            xp2     = prior['x'][ip2]
        
        # Method: linear regression (Python)
        # https://github.com/JoramSoch/MBI/blob/main/Python/cvBMS.py
        if meth[h] == 'LinReg':
            P1  = np.linalg.inv(V1)
            if Mlinreg == 'OLS':
                b_est1 = np.linalg.inv(Y1.T @ Y1) @ Y1.T @ x1
            elif Mlinreg == 'WLS':
                b_est1 = np.linalg.inv(Y1.T @ P1 @ Y1) @ Y1.T @ P1 @ x1
            else:
                b_est1 = np.zeros((v,1))
            xp2 = Y2 @ b_est1

        # Method: support vector regression (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVR.html
        if meth[h] == 'SVR':
            svm1 = skl.svm.SVR(kernel=Ksvm, C=Csvm).fit(Y1, x1)
            xp2  = svm1.predict(Y2)
        
        # Method: random forrest regression (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html
        if meth[h] == 'RFR':
            rf1 = skl.ensemble.RandomForestRegressor(
                n_estimators=Nrf).fit(Y1, x1)
            xp2 = rf1.predict(Y2)
        
        # Method: neural network regression (scikit-learn)
        # https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPRegressor.html
        if meth[h] == 'NNR':
            nn1 = skl.neural_network.MLPRegressor(
                hidden_layer_sizes=Lnn, activation=Ann).fit(Y1, x1)
            xp2 = nn1.predict(Y2)
        
        # store test set predictions
        xp[i2,h]   = xp2
        
    # delete analysis variables
    del mba1, pp2, ip2, B_est1, s2_est1, Si_est1, log_PP, P1, b_est1, svm1, rf1, nn1

# calculate performance
for h in range(M):
    r[h]   = np.corrcoef(xp[:,h], x)[0,1]
    MAE[h] = np.mean(np.absolute(xp[:,h]-x))
    RGA[h] = MBI.calc_RGA(x, xp[:,h], 'RGA')


### Step 3: visualize results #################################################

# open figure
fig   = plt.figure(figsize=(8,9))
ax    = fig.add_subplot(111)
x_off = 2
y_off = 0.98
cols  = [[  0,  32,  96], [  0,   0, 255],
         [192,   0,   0], [255,   0,   0], [255, 192,   0], [255, 255,   0]]

# plot performance
for h in range(M):
    col = tuple([rgb/255 for rgb in cols[h]])
    ax.bar(h+1, RGA[h], width= 0.7, color=col, edgecolor='k', label=meth[h])
ax.plot([(1-1), (M+2)], [1/2, 1/2],     ':k', linewidth=2, label='chance')
ax.plot([x_off+1/2, x_off+1/2], [0, 1], '-k', linewidth=1)
ax.axis([(1-1), (M+2), 0, 1])
ax.set_xticks([])
ax.legend(loc='lower right', fontsize=12)
ax.set_xlabel('classification approach', fontsize=20)
ax.set_ylabel('rank graduation accuracy', fontsize=20)
ax.set_title('Simulation 4', fontsize=24, fontweight='bold')
ax.tick_params(axis='both', labelsize=12)
ax.text(x_off+1/2, y_off, 'generative methods   ',
        fontsize=14, ha='right', va='center')
ax.text(x_off+1/2, y_off, '   discriminative methods',
        fontsize=14, ha='left', va='center')

# enable tight layout
fig.tight_layout()